### DISCUSSION TITLE
sdkman Installed JDKs Run Slow on Macbook Pro M2

### DISCUSSION THREAD

**Discussion by crgardner** [sdkman/sdkman-cli#1306] [Q&A] [answered]
I'm hesitating to file an issue because it may be machine specific. I have the following setup

* Apple M2 Pro
* Sonoma 14.6
* Sdkman recently installed and using
  - Maven 3.9.8
  - Java 21.0.4-tem (also have used 21.0.4-amzn)
  
 I have a Java 21 multi-module maven project. The code base is small. When I run mvn clean package the build takes 1 minute and 16 seconds. Using mvnd doesn't improve it that much. Also, if I manually set JAVA_HOME and PATH to use the sdkman installed JDK, the results are similar.

If, however, I install temurin-21.0.4 another way (via IntelliJ download, specifically) and manually set JAVA_HOME and PATH and run mvn clean package, the build time is about 21 seconds. mvnd completes it in 8 seconds.

I welcome any suggestions for troubleshooting.

**Comment by marc0der** [comment]
Hi @crgardner, we download the latest SDKs (including JDKs) directly from the source and unpack it on your machine. Also, we already manage the `JAVA_HOME` and `PATH` on your half, so you **should not be setting those yourself**.

The only thing that I think could be wrong is that you are not downloading the incorrect JDK variant in SDKMAN. We used the Intel JDKs with Rosetta 2 compatibility in the past when there were no ARM JDKs around. When ARM JDKs became mainstream on Mac, we switched SDKMAN to turn off Rosetta and download the ARM version.

Can you confirm that you are using the latest version of SDKMAN? What does `sdk version` render?

**Comment by crgardner** [comment]
I totally understand about not setting those environment variables myself. I was doing that for troubleshooting only. Meanwhile, here's the version I'm using:

$ sdk version

SDKMAN!
script: 5.18.2
native: 0.4.6

**Reply by crgardner** [reply]
Here's my current setup. I ran with mvnd and am still seeing slow build times (38 seconds). Using the manually downloaded and setup temurin-21.0.4, as mentioned earlier in the thread, along with sdkman installed mvn and mvnd, I see remarkably fast build times (mvnd clocks in just over 7 seconds).

```

$ sdk current java

Using java version 21.0.4-amzn

$ java -version
openjdk version "21.0.4" 2024-07-16 LTS
OpenJDK Runtime Environment Corretto-21.0.4.7.1 (build 21.0.4+7-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.4.7.1 (build 21.0.4+7-LTS, mixed mode, sharing)
```

**Reply by marc0der** [reply]
> I totally understand about not setting those environment variables myself.

Okay great, so with that out of the way, let's try to figure out what the underlying cause is.

Can you let me know what you see when you run the following lines?

```bash
cat $HOME/.sdkman/var/platform

cat $HOME/.sdkman/etc/config
```

It would also be good if you could do a local install of your external JDK into SDKMAN and test it from there:

```bash
sdk install java local /path/to/your/local/jdk
sdk use java local
# now run maven
```

Lastly, do you know if the version of Temurin 21.0.4 you installed is for the Intel or Arm64 architecture?

**Reply by crgardner** [reply]
```
$ cat $HOME/.sdkman/var/platform
darwinx64

$ cat $HOME/.sdkman/etc/config
sdkman_auto_answer=false
sdkman_auto_complete=true
sdkman_auto_env=false
sdkman_beta_channel=false
sdkman_checksum_enable=true
sdkman_colour_enable=true
sdkman_curl_connect_timeout=7
sdkman_curl_max_time=10
sdkman_debug_mode=false
sdkman_insecure_ssl=false
sdkman_selfupdate_feature=true

$ file `which java`
/Users/gardnchr/Library/Java/JavaVirtualMachines/temurin-21.0.4/Contents/Home/bin/java: Mach-O 64-bit executable arm64

$ sdk install java local /Users/gardnchr/Library/Java/JavaVirtualMachines/temurin-21.0.4/Contents/Home
$ sdk use java local

$ mvnd clean package
```

With your suggested configuration, mvnd took 8 seconds, consistent with my manual configuration.

I can try to do a clean install of sdkman if necessary. Incidentally, this was on a new laptop.

**Reply by helpermethod** [reply, marked as answer]
@marc0der 

Shouldn't the platform be sth like

`darwinarm64`?

**Reply by marc0der** [reply]
Yes, exactly that. I wonder if this installation wasn't copied from an Intel Mac at some point.

So, my suspicion was right. SDKMAN is downloading the x86 version of Java (and anything else platform-specific) and running it through the Rosetta 2 adaptive layer.

**Reply by crgardner** [reply]
I removed the previous installation completely and reinstalled. This has solved the problem.

**Reply by marc0der** [reply]
Fantastic news, I'm glad it's working now.
